### PR TITLE
Create a wrapper around QQmlComponent

### PR DESCRIPTION
**Pull request by agentsim**
Support setData and loadUrl methods
Fix SignalCppRepresentation to support classes other than QObject

### CODE REVIEW

File: qmetaobject/src/qtdeclarative.rs [hunk 1]
```diff
@@ -201,6 +202,106 @@ impl Default for QQuickView {
     }
 }
 
+#[repr(u32)]
+#[derive(Clone, Copy)]
+pub enum CompilationMode {
```
Comment (ogoffart) [inline qmetaobject/src/qtdeclarative.rs]: Please add documentation

File: qmetaobject/src/qtdeclarative.rs [hunk 2]
```diff
@@ -201,6 +202,106 @@ impl Default for QQuickView {
...
+    PreferSynchronous,
+    Asynchronous,
+}
+
+#[repr(u32)]
+#[derive(Clone, Copy)]
+pub enum ComponentStatus {
```
Comment (ogoffart) [inline qmetaobject/src/qtdeclarative.rs]: Please add documentation

File: qmetaobject/src/qtdeclarative.rs [hunk 3]
```diff
@@ -201,6 +202,106 @@ impl Default for QQuickView {
...
+    /// Create a QmlComponent using the QmlEngine.
+    pub fn new(engine: &QmlEngine) -> QmlComponent {
+        unsafe {
+            cpp!([engine as "QmlEngineHolder*"] -> QmlComponent as "QQmlComponentHolder" {
+                return QQmlComponentHolder(engine->engine.get());
+            })
+        }
+    }
+
+    pub fn get_cpp_object(&self) -> *mut c_void {
```
Comment (ogoffart) [inline qmetaobject/src/qtdeclarative.rs]: Could also have some documentation.
(Also explain what's the relation with the method in the QObject trait)

File: qmetaobject/src/qtdeclarative.rs [hunk 4]
```diff
@@ -201,6 +202,106 @@ impl Default for QQuickView {
...
+
+    /// Performs QQmlComponent::create
+    pub fn create(&mut self) -> *mut c_void {
+        unsafe { cpp!([self as "QQmlComponentHolder*"] -> *mut c_void as "QObject*" {
+            return self->component->create();
+        })}
+    }
+
+    /// See Qt documentation for QObject::destroyed
+    pub fn status_changed_signal() -> CppSignal<fn(status: ComponentStatus)> {
```
Comment (ogoffart) [inline qmetaobject/src/qtdeclarative.rs]: Documentation needs fix.  
I also would like to see this used in a test, please.

Comment (agentsim) [reply qmetaobject/src/qtdeclarative.rs]: I've copied ```test_future``` as a template for implementing this test. In so doing I've noticed that both my test and ```test_future``` crash on macOS Catalina with Qt 5.13.1 and nightly Rust.

**Comment by ogoffart** [comment]
There are two problems with the tests:

1. The tests `create_component` and `component_status_changed` are run in parallel in different thread and both create a QmlEngine, which itself create a QApplication.  But there can only be one QApplication per object.  In fact there should only be one instance of QmlEngine.  I realize that the API is unsound and QQmlEngine should better panic! if it tries to instentiate a second one.  But the solution for your test is to move them in the `tests` directory, there you can acquire the `lock_for_test()`

2. You must somehow call `quit` in the QmlEngine otherwise it is gonna run forever.

**Comment by ogoffart** [comment]
Thanks for the contributions!

**Pull request merged**